File: lms-studio/schemaTypes/student.ts
import { defineType, defineField } from 'sanity'

export default defineType({
  name: 'student',
  title: 'Student',
  type: 'document',
  fields: [
    defineField({
      name: 'name',
      title: 'Full Name',
      type: 'string',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'email',
      title: 'Email',
      type: 'string',
      validation: (Rule) => Rule.required().email(),
    }),
    defineField({
      name: 'rollNumber',
      title: 'Roll Number',
      type: 'string',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'phone',
      title: 'Phone Number',
      type: 'string',
    }),
    defineField({
      name: 'dateOfBirth',
      title: 'Date of Birth',
      type: 'date',
    }),
    defineField({
      name: 'address',
      title: 'Address',
      type: 'text',
    }),
    defineField({
      name: 'profileImage',
      title: 'Profile Image',
      type: 'image',
      options: {
        hotspot: true,
      },
    }),
    defineField({
      name: 'batch',
      title: 'Batch',
      type: 'reference',
      to: [{ type: 'batch' }],
    }),
    defineField({
      name: 'enrolledCourses',
      title: 'Enrolled Courses',
      type: 'array',
      of: [{ type: 'reference', to: [{ type: 'course' }] }],
    }),
    defineField({
      name: 'isActive',
      title: 'Is Active',
      type: 'boolean',
      initialValue: true,
    }),
  ],
  preview: {
    select: {
      title: 'name',
      subtitle: 'rollNumber',
      media: 'profileImage',
    },
  },
})

File: l_m_s/features/lms/feature_lms/lib/models/student.dart
/// Domain model for a student (Sanity student type).
class Student {
  final String id;
  final String name;
  final String? email;
  final String? rollNumber;
  final String? profileImageUrl;
  final List<String> enrolledCourseIds;
  final String? batchId;
  final bool isActive;

  const Student({
    required this.id,
    required this.name,
    this.email,
    this.rollNumber,
    this.profileImageUrl,
    this.enrolledCourseIds = const [],
    this.batchId,
    this.isActive = true,
  });

  factory Student.fromMap(Map<String, dynamic> map) {
    String? profileUrl;
    final img = map['profileImage'];
    if (img is Map && img['asset'] != null) {
      final asset = img['asset'];
      if (asset is Map && asset['url'] != null) profileUrl = asset['url'] as String?;
    }
    if (profileUrl == null && map['profileImageUrl'] != null) profileUrl = map['profileImageUrl'] as String?;

    List<String> ids = [];
    final enrolled = map['enrolledCourses'];
    if (enrolled is List) {
      for (final e in enrolled) {
        if (e is Map) {
          final ref = e['_id'] ?? e['_ref'];
          if (ref != null) ids.add(ref.toString());
        }
      }
    }
    if (ids.isEmpty && map['enrolledCourseIds'] is List) ids = List<String>.from(map['enrolledCourseIds'] as List);

    return Student(
      id: map['_id'] as String? ?? '',
      name: map['name'] as String? ?? '',
      email: map['email'] as String?,
      rollNumber: map['rollNumber'] as String?,
      profileImageUrl: profileUrl,
      enrolledCourseIds: ids,
      batchId: map['batch'] is Map ? (map['batch']['_ref'] ?? map['batch']['_id']) as String? : map['batchId'] as String?,
      isActive: map['isActive'] as bool? ?? true,
    );
  }
}
